File: assignment 5/c programming/p8.c
//Give a singly linked list of size N. The task is to left-shift the linked list by k nodes, where k is a positive integer smaller than or equal to length of the linked list.
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

void left_shift(struct Node** head_ref, int k) {
    if (*head_ref == NULL || k <= 0) return;

    struct Node* current = *head_ref;
    int length = 1;

    // Find the last node and the length of the list
    while (current->next != NULL) {
        current = current->next;
        length++;
    }

    // If k is greater than length, normalize it
    k = k % length;
    if (k == 0) return;

    // Find the new head after k shifts
    struct Node* new_head = *head_ref;
    struct Node* prev = NULL;

    for (int i = 0; i < k; i++) {
        prev = new_head;
        new_head = new_head->next;
    }

    // Update the links
    current->next = *head_ref;
    *head_ref = new_head;
    prev->next = NULL;
}

int main() {
    struct Node* head = NULL;
    // Create the linked list and populate it
    for (int i = 1; i <= 5; i++) {
        struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
        new_node->data = i;
        new_node->next = head;
        head = new_node;
    }

    int k = 2;  // Number of positions to left shift
    left_shift(&head, k);

    // Print the modified linked list
    struct Node* current = head;
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");

    return 0;
}

File: assignment 5/python/p8.py
#Give a singly linked list of size N. The task is to left-shift the linked list by k nodes, where k is a positive integer smaller than or equal to length of the linked list.
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def add(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

    def show(self):
        current = self.head
        while current:
            print(current.data, end=" -> ")
            current = current.next
        print("None")

    def left_shift(self, k):
        if self.head is None or self.head.next is None or k <= 0:
            return  # No need to shift if list is empty, has only one element, or k is invalid

        # Find the length of the list
        length = 0
        current = self.head
        while current:
            length += 1
            current = current.next

        # Normalize k in case it's greater than length
        k = k % length
        if k == 0:
            return  # No need to shift if k is 0

        # Find the new head after k shifts
        current = self.head
        for _ in range(k - 1):
            current = current.next
        new_head = current.next
        current.next = None

        # Find the last node and connect it to the old head
        last = new_head
        while last and last.next:
            last = last.next
        if last:
            last.next = self.head
        self.head = new_head

# Example usage
my_list = LinkedList()
my_list.add(10)
my_list.add(20)
my_list.add(30)
my_list.show()  # Output: 10 -> 20 -> 30 -> None
my_list.left_shift(2)
my_list.show()  # Output: 30 -> 10 -> 20 -> None
